packages/material-ui/src/Switch/Switch.test.js

```diff
@@ -38,35 +33,58 @@ describe('<Switch />', () => {
-      assert.strictEqual(typeof classes.root, 'string');
-      assert.strictEqual(typeof classes.checked, 'string');
-      assert.strictEqual(typeof classes.disabled, 'string');
+      expect(classes).to.include.all.keys(['root', 'checked', 'disabled']);
     });
   });
 
-  describe('default Switch export', () => {
-    let wrapper;
+  specify('should render an .thumb element inside the .switchBase element', () => {
+    const { container } = render(
+      <Switch classes={{ thumb: 'thumb', switchBase: 'switch-base' }} />,
+    );
 
-    beforeEach(() => {
-      wrapper = shallow(<Switch className="foo" />);
-    });
+    expect(container.querySelector('.switch-base .thumb')).to.be.ok;
+  });
 
-    it('should render SwitchBase with a custom span icon with the thumb class', () => {
-      const switchBase = wrapper.childAt(0);
-      assert.strictEqual(switchBase.type(), SwitchBase);
-      assert.strictEqual(switchBase.props().icon.type, 'span');
-      assert.strictEqual(switchBase.props().icon.props.className, classes.thumb);
-      assert.strictEqual(switchBase.props().checkedIcon.type, 'span');
-      assert.strictEqual(
-        switchBase.props().checkedIcon.props.className,
-        clsx(classes.thumb, classes.thumbChecked),
-      );
-    });
+  it('should render the track as the 2nd child', () => {
+    const {
+      container: { firstChild: root },
+    } = render(<Switch />);
 
-    it('should render the track as the 2nd child', () => {
-      const track = wrapper.childAt(1);
-      assert.strictEqual(track.name(), 'span');
-      assert.strictEqual(track.hasClass(classes.track), true);
-    });
+    expect(root.childNodes[1]).to.have.property('tagName', 'SPAN');
+    expect(root.childNodes[1]).to.have.class(classes.track);
+  });
+
+  it('renders a `role="checkbox"` with the Unechecked state by default', () => {
+    const { getByRole } = render(<Switch />);
+
+    expect(getByRole('checkbox')).to.have.property('checked', false);
+  });
+
+  it('renders a checkbox with the Checked state when checked', () => {
+    const { getByRole } = render(<Switch defaultChecked />);
+
+    expect(getByRole('checkbox')).to.have.property('checked', true);
+  });
+
+  specify('the switch can be disabled', () => {
+    const { getByRole } = render(<Switch disabled />);
+
+    expect(getByRole('checkbox')).to.have.property('disabled', true);
+  });
+
+  specify('the switch can be readonly', () => {
+    const { getByRole } = render(<Switch readOnly />);
+
+    expect(getByRole('checkbox')).to.have.property('readOnly', true);
+  });
+
+  specify('the Checked state changes after change events', () => {
+    const { getByRole } = render(<Switch defaultChecked />);
+
+    // how a user would trigger it
+    getByRole('checkbox').click();
+    fireEvent.change(getByRole('checkbox'), { target: { checked: '' } });
+
+    expect(getByRole('checkbox')).to.have.property('checked', false);
```

Comment:
Should we assert
```jsx
expect(getByRole('checkbox')).to.have.property('checked', true);
```

before the click event? We have a test for it higher in the file, but I'm wondering if making the assertion leave no room for doubt wouldn't be better.

Reply:
As you've mentioned the test above covers this. It's an implicit requirement for tests that an action allows the assertion. You need another test for the before assertion. Otherwise you end up having to repeat every assertion in every test essentially duplicating either your test suite or what your test does.

Reply:
Ok, the reduction of duplicated assertions makes sense. As long as it doesn't harm the potential to catch regression, it should be fine 👌

Reply:
> It's an implicit requirement for tests that an action allows the assertion. 

@eps1lon Hi, sorry I was wondering if you could expand upon this? I'm still getting used to testing and not sure what you mean by this! (Sorry if this isn't the place to ask, but didn't know where to best ask).

Reply:
@forbesd7 I think that it's meant to say that if a previous test covers the outcome of the same action, it's OK-ish to omit the assertion, it can avoid duplication.

Reply:
I think I meant that you always assume that a test fails fails if you remove the "action". If you change you "unit" so that removing the action still passes this specific test then another test would fail. See ["arrange - act- assert" testing pattern](https://wiki.c2.com/?ArrangeActAssert)

Reply:
Ah okay that makes sense! Thank you @oliviertassinari @eps1lon for the insight and resources! :)